intermediate_source/imbalanced_data_tutorial.py

```diff
@@ -0,0 +1,486 @@
+        loss = criterion(output, target)
+        # Backward pass
+        loss.backward()
+        # Weight update
+        optimizer.step()
+
+        # print statistics
+        running_loss += loss.item()
+        if batch_idx % 100 == 0:  # print every 100 mini-batches
+            print('[{}, {}] loss: {:.3f}'.format(epoch + 1, batch_idx + 1,
+                                                 running_loss / 100))
+            running_loss = 0.0
+
+
+def evaluate():
+    model.eval()
+    correct = 0
+    total = 0
+    confusion_matrix = torch.zeros(nb_classes, nb_classes)
+    with torch.no_grad():
+        for data, target in test_loader:
+            output = model(data)
+            _, preds = torch.max(output, 1)
+
+            correct += (preds == target).sum().item()
+            total += target.size(0)
+            for t, p in zip(target, preds):
+                confusion_matrix[t, p] += 1
+
+    # Calculate global accuracy
+    accuracy = 100 * correct / total
+
+    # Calculate class accuracies
+    class_correct = confusion_matrix.diag()
+    class_total = confusion_matrix.sum(1)
+    class_accuracies = class_correct / class_total
+
+    # Normalize confusion matrix
+    for i in range(nb_classes):
+        confusion_matrix[i] = confusion_matrix[i] / confusion_matrix[i].sum()
+
+    # Print statistics
+    print('Accuracy of the model {:.3f}%'.format(accuracy))
+    for i in range(nb_classes):
+        print('Accuracy for {}: {:.3f}%'.format(
+            class_names[i], 100 * class_accuracies[i]))
+    print('Mean per class accuracy: {:.3f}%'.format(
+        100 * class_accuracies.mean()))
+
+    # Plot confusion matrix
+    f = plt.figure()
+    ax = f.add_subplot(111)
+    cax = ax.imshow(confusion_matrix.numpy(), interpolation='nearest')
+    f.colorbar(cax)
+    plt.xticks(range(len(class_names)), class_names, rotation=90)
+    plt.yticks(range(len(class_names)), class_names)
+
+
+###############################################################################
+# Let's have a look at the performance of our model on the test dataset.
+# It's also interesting to see the confusion matrix, where each row represents
+# the target instances while each column represents the predicted instances.
+
+for epoch in range(25):
+    train(epoch)
+evaluate()
+
+###############################################################################
+# As expected the network performs quite good and the accuracies for each class
+# are very similar.
+#
+# Creating an imbalanced dataset
+# =================================
+#
+# Now let's simulate an imbalanced dataset.
+# We will resample the training and test dataset, so that some
+# classes will be undersampled. Let's reduce the amount of instances for the
+# first 5 classes by 90% leaving only the first 10% in the dataset.
+# The last 5 classes will keep their samples.
+
+# Create class proportions
+imbal_class_prop = imbal_class_prop = np.hstack(([0.1] * 5, [1.0] * 5))
```

Comment:
I think rather than relying on the original dataset's code, it's better to create a new Dataset class which creates something like `self.idxs` and in ``__getitem`` can return ``self.orig_dataset[self.idxs[i]]``. You can use this dataset for both train and test datasets.

Reply:
Current code to do this is a bit opaque. It's always best to have no code duplication.

Reply:
Yeah, I've had the same feeling. I created a ``Dataset`` with a function to resample the data.
However, since I need the targets and class counts for the weights, I've added a helper class to get these members. It's better, but maybe still not really clean.